cmd/protoc-gen-gorums/dev/strictordering.go

```diff
@@ -0,0 +1,165 @@
+package dev
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync"
+	atomic "sync/atomic"
+	"time"
+
+	"github.com/relab/gorums"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/codes"
+	status "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+type strictOrderingResult struct {
+	nid   uint32
+	reply *anypb.Any
+	err   error
+}
+
+type requestHandler func(*gorums.Message) *gorums.Message
+
+type strictOrderingManager struct {
+	msgID    uint64
+	recvQ    map[uint64]chan *strictOrderingResult
+	recvQMut sync.RWMutex
+}
+
+func newStrictOrderingManager() *strictOrderingManager {
+	return &strictOrderingManager{
+		recvQ: make(map[uint64]chan *strictOrderingResult),
+	}
+}
+
+func (m *strictOrderingManager) createStream(node *Node, backoff backoff.Config) *strictOrderingStream {
+	return &strictOrderingStream{
+		node:      node,
+		nextMsgID: m.nextMsgID,
+		backoff:   backoff,
+		rand:      rand.New(rand.NewSource(time.Now().UnixNano())),
+		sendQ:     make(chan *gorums.Message),
+		recvQ:     m.recvQ,
+		recvQMut:  &m.recvQMut,
+	}
+}
+
+func (m *strictOrderingManager) nextMsgID() uint64 {
+	return atomic.AddUint64(&m.msgID, 1)
+}
+
+type strictOrderingStream struct {
+	node         *Node         // needed for ID and setLastError
+	nextMsgID    func() uint64 // needed for Node RPC methods
+	backoff      backoff.Config
+	rand         *rand.Rand
+	gorumsClient gorums.GorumsClient
+	gorumsStream gorums.Gorums_StrictOrderingClient
+	streamMut    sync.RWMutex
+	streamBroken bool
+	sendQ        chan *gorums.Message
+	recvQ        map[uint64]chan *strictOrderingResult
+	recvQMut     *sync.RWMutex
+}
+
+func (s *strictOrderingStream) connect(ctx context.Context, conn *grpc.ClientConn) error {
+	var err error
+	s.gorumsClient = gorums.NewGorumsClient(conn)
+	s.gorumsStream, err = s.gorumsClient.StrictOrdering(ctx)
+	if err != nil {
+		return err
+	}
+	go s.sendMsgs()
+	go s.recvMsgs(ctx)
+	return nil
+}
+
+func (s *strictOrderingStream) sendMsgs() {
+	for req := range s.sendQ {
+		// return error if stream is broken
+		if s.streamBroken {
+			err := status.Errorf(codes.Unavailable, "stream is down")
+			s.recvQMut.RLock()
```

Comment:
Should this err be recorded?
```go
	s.node.setLastErr(err)
```

Reply:
I don't think it's necessary to record this error since it doesn't add any information about why the node stream was closed. Maybe we could return `node.LastErr()` instead of this generic error to give specific information about why the node errored? If `node.SetLastErr()` is called by `reconnectStream`, then it will probably contain the error that caused the stream to fail. Of course there is a chance that some unrelated error will have taken its place.